### PR TITLE
Add dashlane adapter

### PR DESCRIPTION
**Pull request by skinnyfists**
This PR adds an adapter for the Dashlane CLI, as per #1761, as well as tests for it.

Dashlane supports both _password_ and _secret_ as models for storing credentials, but secrets might be enterprise only (I don't have access to them), hence the implementation with indifferent access to both. This behaviour could be changed if not ideal.

`account`'s only purpose in this adapter is to verify that we are logged in with the right account if already logged in, so it might not be necessary if being logged in to the wrong accoung is out of scope as an edge case.

### CODE REVIEW

File: test/secrets/dashlane_adapter_test.rb [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+    end
+  end
+
+  test "fetch with failed login" do
+    stub_cli_installed(true)
+    stub_status(false)
+    stub_login(false)
+
+    assert_raises(RuntimeError, "Failed to login to or unlock Dashlane") do
+      @adapter.fetch([ "SECRET1, PASSWORD1" ], account: account)
```
[the remark this answers is not in the document]
Comment (skinnyfists) [reply test/secrets/dashlane_adapter_test.rb]: fixed in cdabc5763c6a7d3ab40fd5d3518035f89a44922f

File: test/secrets/dashlane_adapter_test.rb [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+  end
+
+  test "fetch with missing entries" do
+    setup_logged_in
+    stub_dashlane_password(found: false, secrets: secrets)
+    stub_dashlane_secret(found: false, secrets: secrets)
+
+    assert_raises RuntimeError, "Could not find #{secrets.keys.join(", ")} in Dashlane passwords or secrets" do
+      @adapter.fetch(secrets.keys, account: account)
+    end
```
[the remark this answers is not in the document]
Comment (skinnyfists) [reply test/secrets/dashlane_adapter_test.rb:57]: fixed in 85e465ff93e40e0941783ad510dff1916b26d74d

File: test/secrets/dashlane_adapter_test.rb [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+  setup do
+    @adapter = Kamal::Secrets::Adapters::Dashlane.new
+  end
+
+  test "fetch without CLI installed" do
+    stub_cli_installed(false)
+
+    assert_raises(RuntimeError, "Dashlane CLI is not installed") do
+      run_command("fetch", *secrets.keys)
+    end
```
[the remark this answers is not in the document]
Comment (skinnyfists) [reply test/secrets/dashlane_adapter_test.rb:13]: fixed in 85e465ff93e40e0941783ad510dff1916b26d74d

File: test/secrets/dashlane_adapter_test.rb [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+    stub_cli_installed(true)
+    stub_status(false)
+    stub_login(true)
+    stub_dashlane_password(found: true, secrets: secrets)
+    stub_dashlane_secret(found: true, secrets: secrets)
+
+    result = @adapter.fetch(secrets.keys, account: account)
+
+    assert_equal(result.sort, secrets.sort)
+  end
```
[the remark this answers is not in the document]
Comment (skinnyfists) [reply test/secrets/dashlane_adapter_test.rb]: fixed in fc276b5a87d6bf548e5f80e8e12023ceef4c0d15

File: test/secrets/dashlane_adapter_test.rb [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+        Login: #{account}
+        Locked: no
+        STATUS
+      else
+        stub_ticks_with("dcli status 2> /dev/null", succeed: false)
+      end
+    end
+
+    def stub_login(success)
+      @adapter.stubs(:system).with { "dcli sync 2> /dev/null" && (success ? `true` : `false`) }
```
[the remark this answers is not in the document]
Comment (skinnyfists) [reply test/secrets/dashlane_adapter_test.rb]: fixed in d17adc889c9ea7e01e3566fdf55164f1d456c50e

File: test/secrets/dashlane_adapter_test.rb [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+
+    def run_command(*command)
+      stdouted do
+        Kamal::Cli::Secrets.start \
+          [ *command,
+            "-c", "test/fixtures/deploy_with_accessories.yml",
+            "--adapter", "dashlane",
+            "--account", "email@example.com" ]
+      end
+      Kamal::Secrets::Adapters::Dashlane.new
```
[the remark this answers is not in the document]
Comment (skinnyfists) [reply test/secrets/dashlane_adapter_test.rb]: woops, forgot about that. removed in d17adc889c9ea7e01e3566fdf55164f1d456c50e

File: lib/kamal/secrets/adapters/dashlane.rb [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+      Locked: no
+      STATUS
+      `dcli status 2> /dev/null` == status
+    end
+
+    def fetch_secrets(secrets, from:, account:, session:)
+      shell_secrets_string = secrets.map(&:shellescape).join(" ")
+      dashlane_passwords = JSON.parse(`dcli password #{shell_secrets_string} -o json`)
+      dashlane_secrets = JSON.parse(`dcli secret #{shell_secrets_string} -o json`)
+
```
[the remark this answers is not in the document]
Comment (skinnyfists) [reply lib/kamal/secrets/adapters/dashlane.rb:34]: Added a safety check in c2c65f80a6194c413d1147d5aa797d760b466f34

File: lib/kamal/secrets/adapters/dashlane.rb [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+    def logged_in?(account)
+      status = <<~STATUS
+      Logged in: yes
+      Login: #{account}
+      Locked: no
+      STATUS
+      `dcli status 2> /dev/null` == status
+    end
+
+    def fetch_secrets(secrets, from:, account:, session:)
```
[the remark this answers is not in the document]
Comment (skinnyfists) [reply lib/kamal/secrets/adapters/dashlane.rb:26]: Unless I missed something, Dashlane doesn't seem to have vaults and that's what the `from` argument seems to be used for in other adapters. I added a guard clause that raises an error in c2c65f80a6194c413d1147d5aa797d760b466f34 but will await maintainer input on expected behaviour

File: test/secrets/dashlane_adapter_test.rb [hunk 9]
```diff
@@ -0,0 +1,144 @@
...
+
+    error = assert_raises RuntimeError do
+      @adapter.fetch(secrets.keys, account: account)
+    end
+    assert_equal "Could not find #{secrets.keys.join(", ")} in Dashlane passwords or secrets", error.message
+  end
+
+  test "fetch dashlane passwords and no secrets" do
+    only_passwords = secrets.select { |k| k.match? "PASSWORD" }
+    setup_logged_in
```
[the remark this answers is not in the document]
Comment (skinnyfists) [reply test/secrets/dashlane_adapter_test.rb]: Actually, `String#match?` converts the passed pattern to a regexp so this works as expected. I changed it to `include?` nonetheless because it seems more idiomatic
f5292f7cc602e8f03caad0d2c1d24e872a4ea3b5

File: lib/kamal/secrets/adapters/dashlane.rb [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+      raise ArgumentError, "Dashlane adapter does not support the --from option" if from.present?
+
+      shell_secrets_string = secrets.map(&:shellescape).join(" ")
+      dashlane_passwords = `dcli password #{shell_secrets_string} -o json`
+      raise RuntimeError, "Could not read #{secrets} from Dashlane passwords" unless $?.success?
+      dashlane_secrets = `dcli secret #{shell_secrets_string} -o json`
+      raise RuntimeError, "Could not read #{secrets} from Dashlane secrets" unless $?.success?
```
[the remark this answers is not in the document]
Comment (skinnyfists) [reply lib/kamal/secrets/adapters/dashlane.rb:33]: That's not how it works. Secrets might not be accessible for creation in the web interface but the cli always works, and will just return an empty array if you haven't created any secrets.
